pkgs/development/python-modules/openai-triton/default.nix

```diff
@@ -0,0 +1,231 @@
+    substituteInPlace bin/CMakeLists.txt \
+      --replace "add_subdirectory(FileCheck)" ""
+
+    rm cmake/FindLLVM.cmake
+  ''
+  +
+  (
+    let
+      # Bash was getting weird without linting,
+      # but basically upstream contains [cc, ..., "-lcuda", ...]
+      # and we replace it with [..., "-lcuda", "-L/run/opengl-driver/lib", "-L$stubs", ...]
+      old = [ "-lcuda" ];
+      new = [ "-lcuda" "-L${addOpenGLRunpath.driverLink}" "-L${cuda_cudart}/lib/stubs/" ];
+
+      quote = x: ''"${x}"'';
+      oldStr = lib.concatMapStringsSep ", " quote old;
+      newStr = lib.concatMapStringsSep ", " quote new;
+    in
+    ''
+      substituteInPlace python/triton/compiler.py \
+        --replace '${oldStr}' '${newStr}'
+    ''
+  )
+  # Triton seems to be looking up cuda.h
+  + ''
+    sed -i 's|cu_include_dir = os.path.join.*$|cu_include_dir = "${cuda_cudart}/include"|' python/triton/compiler.py
+  '';
+
+  nativeBuildInputs = [
+    cmake
+    (llvmPackages.llvm.override {
+      llvmTargetsToBuild = [ "NATIVE" "NVPTX" ];
+      # Upstream CI sets these too:
+      # targetProjects = [ "mlir" ];
+      extraCMakeFlags = [
+        "-DLLVM_INSTALL_UTILS=ON"
+      ];
+    })
+    llvmPackages.mlir
+    lit
+    pythonRelaxDepsHook
+  ];
+
+  buildInputs = [
+    pybind11
+    gtest
+    zlib
+    ncurses
+    libxml2.dev
+  ];
+
+  propagatedBuildInputs = [
+    filelock
+  ];
+
+  # Avoid GLIBCXX mismatch with other cuda-enabled python packages
+  preConfigure =
+    ''
+      export CC="${backendStdenv.cc}/bin/cc";
+      export CXX="${backendStdenv.cc}/bin/c++";
+    ''
+    # Upstream's setup.py tries to write cache somewhere in ~/
+    + ''
+      export HOME=$PWD
+    ''
+    # Upstream's github actions patch setup.cfg to write base-dir. May be redundant
+    + ''
+      echo "" >> python/setup.cfg
+      echo "[build_ext]" >> python/setup.cfg
+      echo "base-dir=$PWD" >> python/setup.cfg
+    ''
+    # The rest (including buildPhase) is relative to ./python/
+    + ''
+      cd python/
+    ''
+    # Work around download_and_copy_ptxas()
+    + ''
+      dst_cuda="$PWD/triton/third_party/cuda/bin"
+      mkdir -p "$dst_cuda"
+      ln -s "${ptxas}" "$dst_cuda/"
+    '';
+
+  # CMake is ran by setup.py instead
+  dontUseCmakeConfigure = true;
+  cmakeFlags = [
+    "-DMLIR_DIR=${llvmPackages.mlir}/lib/cmake/mlir"
+
+    # TODO: Probably redundant
+    "-DLLVM_BUILD_LLVM_DYLIB=ON"
+    # "-DLLVM_LINK_LLVM_DYLIB=ON"
+  ];
+
+  postFixup =
+    let
+      ptxasDestination = "$out/${python.sitePackages}/triton/third_party/cuda/bin/ptxas";
+    in
+    # Setuptools (?) strips runpath and +x flags. Let's just restore the symlink
+    ''
+      rm -f ${ptxasDestination}
+      ln -s ${ptxas} ${ptxasDestination}
+    '';
+
+  checkInputs = [
+    cmake # ctest
+  ];
+  dontUseSetuptoolsCheck = true;
+  checkPhase =
+    # Requires torch (circular dependency) and probably needs GPUs:
+    # ''
+    #   (cd test/unit/ ; ${pytest}/bin/pytest)
```

Comment:
Is there a way to do the same with `pytestCheckHook`? I think the point of changing the directory is that `./triton` shouldn't exist so pytest would pick up the installed `triton` from `PYTHONPATH`

Reply:
Add the cd to preCheck.

Reply:
Updated `preCheck`, but cannot confirm if `pytestCheckHook` works because waiting for upstream to resolve the circular dependency issue. Next release